src/lib.rs

```diff
@@ -401,78 +397,79 @@ impl<'hedwig, T, P> PublishBatch<'hedwig, T, P> {
-            }
-            current_topic = topic;
-            current_batch.push(message)
-        }
-        if !current_batch.is_empty() {
-            futures_unordered.push(make_job(current_topic, current_batch));
+        let mut rest = &self.messages[..];
+        // And group them into batches
+        let mut index = 0;
+        while let Some((topic, _)) = rest.first() {
+            let partition_point = partition_point(rest, |(k, _)| k == topic);
+            let (l, r) = rest.split_at(partition_point);
+            rest = r;
+            debug_assert!(l.iter().all(|(k, _)| k == topic));
+            debug_assert!(r.iter().all(|(k, _)| k != topic));
+            let stream = publisher
+                .publish(topic, l.iter().map(|(_, v)| v))
+                .enumerate()
+                .map(move |(i, v)| (i + index, v));
+            index += l.len();
+            select_all.push(stream);
         }
 
-        let mut errors = Vec::new();
-        // Extract the results from all the futures
-        while let (Some(result), stream) = futures_unordered.into_future().await {
-            match result {
-                (_, PublisherResult::Success(ids)) => message_ids.extend(ids),
-                (topic, PublisherResult::OneError(err, failed_msgs)) => {
-                    self.messages
-                        .extend(failed_msgs.into_iter().map(|m| (topic, m)));
-                    errors.push(err);
-                }
-                (topic, PublisherResult::PerMessage(vec)) => {
-                    for message in vec {
-                        match message {
-                            Ok(id) => message_ids.push(id),
-                            Err((err, failed_msg)) => {
-                                self.messages.push((topic, failed_msg));
-                                errors.push(err);
-                            }
-                        }
-                    }
-                }
+        // Await for all of the results...
+        let mut remove = vec![false; self.messages.len()];
+        while let (Some((index, result)), stream) = select_all.into_future().await {
+            remove[index] = result.is_ok();
+            match result_callback(&self.messages[index].1, result) {
+                PublishCallbackAction::Discard => remove[index] = true,
+                PublishCallbackAction::Retry => (),
             }
-            futures_unordered = stream;
+            select_all = stream;
         }
 
-        if let Some(first_error) = errors.pop() {
-            Err(if errors.is_empty() {
-                Error::Publisher(first_error)
-            } else {
-                Error::PublisherMultiple(first_error, errors)
-            })
-        } else {
-            Ok(message_ids)
+        // And remove the messages that have succeeded or have been discarded.
+        {
+            let mut retain_index = 0;
+            self.messages.retain(|_| {
```

Comment:
this surprised me, that a user has to call `publish` again if there were any errors -- my impression from the docs was that `publish` would internally retry any failures that a user marked as "retry".

The docs should probably add a sentence describing the details of retrying (they must check `is_empty` is guess?).